### ISSUE TITLE
Change representation of bytes in text format to something like Base64

### ISSUE THREAD

**Issue by lorenz** [protocolbuffers/protobuf#8598]
**What language does this apply to?**
This is a change to the text encoding of protobuf, not the IDL.

**Describe the problem you are trying to solve.**
The protobuf text format produces output which is hard to read and process for binary data (bytes data type).
Example (I omitted all non-bytes fields):
```prototxt
session {
  session_hash: "N\201\310\345\314\261\364!\310I\200\346\311\376\256|\360\241\365\2508^1`\230\252\265\314\355Nmn"
  nonce: "\203\022@\302\302\264\363|\237H\005Y7\273=\347"
  user_session {
  # ....
    id: "V^]\362\237\3507r>5\026\267\274ea\254"
  }
}
```
Representing arbitrary binary data as ASCII text with escape sequences is IMO unintuitive and very verbose. For example the `session_hash` is 97 characters wide in the standard protobuf text encoding while plain Base64 would take 43 characters which is less than half. The used escape sequences also only work on a small subset of programming languages (for example Go and Rust don't like binary data in strings) and tools, whereas something like Base64 works pretty much everywhere.
Depending on how the new representation is going to look it will also make it possible to easily distinguish between strings and binary data in the protobuf text encoding.
**Describe the solution you'd like**
Something along the lines of
```prototxt
session {
  session_hash: ToHI5cyx9CHISYDmyf6ufPCh9ag4XjFgmKq1zO1ObW4
  nonce: gxJAwsK083yfSAVZN7s95w
  user_session {
  # ....
    id: Vl5d8p/oN3I+NRa3vGVhrA
  }
}
```
I'm not married to that format, but it would have the advantage of being backwards-compatible (note the missing quotes) and is very compact. The only disadvantage this has is it is possible to construct base64 strings which look entirely like integers which might be confusing if you don't have the IDL files handy. It might be preferable to have some form of marker character to signify a binary value, but I haven't found a good one yet.

**Describe alternatives you've considered**
Another encoding could be used, but hexadecimal is verbose and Ascii85 hard to parse as it uses a lot of characters used as delimiters in normal prototext for a very moderate gain in efficiency.

**Comment by haberman** [comment]
Our JSON format uses base64 already, so there is some precedent there. Such a change would require coordination across languages, since several languages implement text format parsers.

Unfortunately the syntax you propose would be non-trivial to implement. The text format parser in C++ [shares a lexer with `.proto` files](https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/io/tokenizer.h). This shared tokenizer doesn't accept unquoted bare strings like this. It would almost parse as an identifier, but `+` and `/` would be problematic, as would a base64 string starting with a digit.

More technically feasible within this framework could be something like:

```
session {
  session_hash: b64 "ToHI5cyx9CHISYDmyf6ufPCh9ag4XjFgmKq1zO1ObW4"
  nonce: b64 "gxJAwsK083yfSAVZN7s95w"
  user_session {
  # ....
    id: b64 "Vl5d8p/oN3I+NRa3vGVhrA"
  }
}
```
